sql/src/main/java/io/crate/operation/collect/files/CSVLineParser.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.collect.files;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvParser;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class CSVLineParser {
+
+    private static CsvMapper csvMapper = new CsvMapper();
+    private static List<Object> keyList;
+    private static ObjectMapper objectMapper = new ObjectMapper();
+
+    public void parseHeader(String header) throws IOException {
+        keyList = csvMapper
+            .enable(CsvParser.Feature.TRIM_SPACES)
+            .readerWithSchemaFor(String.class)
+            .readValues(header.getBytes(StandardCharsets.UTF_8))
+            .readAll();
+
+        Set<Object> keySet  = new HashSet<>(keyList);
+        if (keySet.contains("")) {
+            keySet.removeAll(Collections.singleton(""));
+        }
+
+        if (keySet.size() != keyList.size()) {
+            throw new IllegalArgumentException("Invalid keys!");
+        }
+    }
+
+    public byte[] parse(String row) throws IOException {
+        List<Object> recordList = csvMapper
+            .enable(CsvParser.Feature.TRIM_SPACES)
+            .readerWithTypedSchemaFor(String.class)
+            .readValues(row.getBytes(StandardCharsets.UTF_8))
+            .readAll();
+
+        return convertCSVToMap(recordList);
+    }
+
+    private byte[] convertCSVToMap(List<Object> recordList) throws JsonProcessingException {
```

Comment:
In general, this is quite expensive because it reads all the objects first and then generates the JSON. Would be nice if it were possible to generate the JSON directly from the CSV without creating objects.

Reply:
Basically we would read directly from the bytes, starting a new field once we hit the UTF-8 byte representation of `,`. Of course we would have to take into account quoting of commas and special characters like quotes. Probably too much work to implement. So it's fine.

I'd take a look if Jackson has something like this.

Reply:
```
CsvMapper mapper = new CsvMapper();
// important: we need "array wrapping" (see next section) here:
mapper.enable(CsvParser.Feature.WRAP_AS_ARRAY);
File csvFile = new File("input.csv"); // or from String, URL etc
MappingIterator<Byte[]> it = mapper.readerFor(Byte[].class).readValues(csvFile);
while (it.hasNext()) {
  Byte[] row = it.next();
  // and voila, column values in an array. Works with Lists as well
```

Taken from https://github.com/FasterXML/jackson-dataformats-text/tree/master/csv Just changed the String[] to Byte[]

Wouldn't this give us a byte array directly?

Reply:
Sorry, not sure that makes sense because we have to generate JSON anyways in the end. Forget my comment :)